README.md
# records

An implementation of a simple person-records service.

## Running

### Part I: Core

    $ lein with-profile main-core run random.*

Loads 300 records, 100 each in formats pipe-separated, csv,
space-separated; then prints them out, sorted by various criteria.

### Part II: REST API

    $ lein with-profile main-rest run random.*

Loads the same 300 records, and services REST endpoints on port 3000.

## Testing

### Part I: Core

    $ lein with-profile main-core test

### Part II: REST API

    $ lein with-profile main-rest run random.*
    $ lein with-profile main-rest test

This runs REST API tests against a running REST server.

## Directions for improvement

- Run the integration tests such that after a POST, we have
  original number of records + 1
 


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2021-05-16
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2021-05-16
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/records/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/records/compare/0.1.0...0.1.1


test/records/core_test.clj
(ns records.core-test
  (:require [clojure.test :refer :all]
            [clojure.java.io :as io]
            [records.core :refer :all]
	    [talltale.core :as tt]))

(defn generate-record []
  (let [p (tt/person)
        c (tt/color)]
    (->Record (:last-name p)
              (:first-name p)
	      (:email p)
	      c
	      (:date-of-birth p))))

(defn generate-record-file [filename delimiter count]
  (with-open [f (io/writer filename)]
    (doall
      (repeatedly count
        #(.write f
	         (str (record->line (generate-record) delimiter) "\n"))))))

(deftest split-record-csv
  (testing "successful CSV splitting"
    (is (= ["Doe" "John" " jdoe@example.com  " "Blue" " 1970-02-28 "]
           (record-split
	     "Doe,John, jdoe@example.com  ,Blue, 1970-02-28 "
	     #",")))))

(deftest split-record-pipe
  (testing "successful pipe splitting"
    (is (= ["Doe " " John" " jdoe@example.com  " "Blue" "1970-02-28  "]
           (record-split
	     "Doe | John| jdoe@example.com  |Blue|1970-02-28  "
	     #"[|]")))))

(deftest split-record-spaces
  (testing "successful space splitting"
    (is (= ["Doe" "John" "jdoe@example.com" "Blue" "1970-02-28  "]
           (record-split
	     "Doe  John   jdoe@example.com   Blue\t\t1970-02-28  "
	     #"\s+")))))

(deftest normalizing-legit-email-addr
  (testing "successful email normalization"
    (is (= "john.doe@example.com"
           (normalize-email "  John.Doe@example.COM ")))))

(deftest list-of-legit-email-addr
  (testing "takes only the first addr"
    (is (= "john.doe@example.com"
           (normalize-email "John.doe@example.com, jdoe@gmail.com")))))

(deftest illegitimate-email-addr
  (testing "should reject non-parseable email addr"
    (is (thrown?
         javax.mail.internet.AddressException
	 (normalize-email "asdfsadfasdfsadf@!#!SADDASDAS")))))

(deftest normalizing-color
  (testing "should not regard casing"
    (is (= "blue" (normalize-favorite-color " BlUE   ")))))

(deftest legitimate-iso-birthdate
  (testing "when everything is as expected"
    (is (= (java.time.LocalDate/parse "1970-04-21")
           (normalize-birthdate "1970-04-21")))))

(deftest illegitimate-iso-birthdate
  (testing "when date cannot be parsed"
    (is (thrown?
         java.time.format.DateTimeParseException
	 (normalize-birthdate "1970-04-**")))))

(deftest csv-line->record
  (testing "successful CSV line to record"
    (is (= (->Record
             "Doe"
	     "John"
	     "jdoe@example.com"
	     "blue"
	     (java.time.LocalDate/parse "1970-02-28"))
           (line->record
	     "Doe,John, jdoe@example.com  ,Blue, 1970-02-28 "
	     #",")))))

(deftest pipe-line->record
  (testing "successful pipe line to record"
    (is (= (->Record
             "Doe"
	     "John"
	     "jdoe@example.com"
	     "blue"
	     (java.time.LocalDate/parse "1970-02-28"))
           (line->record
	     "Doe | John| jdoe@example.com  |Blue|1970-02-28  "
	     #"[|]")))))

(deftest spaces-line->record
  (testing "successful space line to record"
    (is (= (->Record
             "Doe"
	     "John"
	     "jdoe@example.com"
	     "blue"
	     (java.time.LocalDate/parse "1970-02-28"))
           (line->record
	     "Doe  John   jdoe@example.com   Blue\t\t1970-02-28  "
	     #"\s+")))))



test/records/rest_test.clj
(ns records.rest-test
  (:require [clojure.test :refer :all]
            [org.httpkit.client :as http]
	    [jsonista.core :as json]))

(deftest get-records-ordered-by-color
  (testing "GET /records/color"
    (let [{:keys [status headers body error] :as resp}
          @(http/get "http://localhost:3000/records/color")
	  cooked (first (json/read-value body))]
      (is (= status 200))
      (is (= 300 (count cooked)))
      (is (= (sort-by :color cooked) cooked)))))

(deftest get-records-ordered-by-name
  (testing "GET /records/name"
    (let [{:keys [status headers body error] :as resp}
          @(http/get "http://localhost:3000/records/name")
	  cooked (first (json/read-value body))]
      (is (= status 200))
      (is (= 300 (count cooked)))
      (is (= (sort-by (juxt :last-name :first-name) cooked) cooked)))))

(deftest get-records-ordered-by-birthdate
  (testing "GET /records/birthdate"
    (let [{:keys [status headers body error] :as resp}
          @(http/get "http://localhost:3000/records/birthdate")
	  cooked (first (json/read-value body))]
      (is (= status 200))
      (is (= 300 (count cooked)))
      (is (= (sort-by :birthdate cooked) cooked)))))

(deftest post-a-text-record
  (testing "POST /records"
    (let [p-resp
            @(http/post "http://localhost:3000/records"
              {:headers {"Content-type" "text/plain"}
               :query-params {"type" "txt"}
               :body "9000 HAL hal@hal.com pink 1992-01-12"})
	  cooked-post (json/read-value (:body p-resp))]
        (is (= (:status p-resp) 200))
        (is (= "HAL" (get cooked-post "first-name"))))))


src/records/core.clj
(ns records.core
  (:gen-class)
  (:require [clojure.string :as str]
            [clojure.java.io :as io]
	    [java-time :as time])
  (:import (javax.mail.internet InternetAddress)
           (java.time LocalDate)))

(defrecord Record
  [last-name first-name email favorite-color birthdate])

(defn normalize-email [addr]
  (->> addr
    (str/trim)
    (InternetAddress/parse)
    (first)
    (str/lower-case)))

(defn normalize-favorite-color [color]
  (str/lower-case (str/trim color)))

(defn normalize-birthdate [birthdate]
  (->> birthdate
    (str/trim)
    (LocalDate/parse)))
    
(def normalizers
  [str/trim
   str/trim
   normalize-email
   normalize-favorite-color
   normalize-birthdate])
   
(defn record-split [raw delimiter]
  (str/split raw delimiter 5))

(defn line->record [line delimiter]
  (->> (record-split line delimiter)
    (map #(%1 %2) normalizers)
    (apply ->Record)))

(defn record->line [record delimiter]
  (->> (map #(%1 record)
            [:last-name :first-name :email :favorite-color :birthdate])
       (str/join delimiter)))

(defn record->display [record delimiter]
  (->> (map #(%1 record)
            [:last-name
	     :first-name
	     :email
	     :favorite-color
	     #(time/format "M/d/yyyy" (:birthdate %1))])
       (str/join delimiter)))

;; Thanks to https://www.rosettacode.org/wiki/Extract_file_extension
(defn file-extension [s]
  (second (re-find #"\.([a-zA-Z0-9]+)$" s)))

(def suffix->delimiter {:txt #"\s+" :csv #"," :pipe #"[|]"})

(defn file->records [filename]
  (let [delim (suffix->delimiter (keyword (file-extension filename)))]
    (with-open [f (io/reader filename)]
      (map #(line->record %1 delim) (doall (line-seq f))))))

(defn -main
  "Read files named in command line args, sort and print"
  [& args]
  (let [recs
         (mapcat file->records args)
        color-then-last-ascending
	 (sort-by (juxt :favorite-color :last-name) recs)
	birthdate-ascending
	 (sort-by :birthdate recs)
	last-descending
	 (reverse (sort-by :last-name recs))]
    (println "Output 1 - by favorite color then last name ascending:")
    (println "=====")
    (doseq [r color-then-last-ascending] (println (record->display r ",")))
    (println)
    (println "Output 2 - by birth date ascending:")
    (println "=====")
    (doseq [r birthdate-ascending] (println (record->display r ",")))
    (println)
    (println "Output 3 - by last name descending:")
    (println "=====")
    (doseq [r last-descending] (println (record->display r ",")))))

src/records/rest.clj
(ns records.rest
  (:gen-class)
  (:require [records.core :as rec]
            [org.httpkit.server :as server]
            [compojure.core :refer :all]
            [compojure.route :as route]
            [ring.middleware.defaults :refer :all]
            [clojure.pprint :as pp]
            [clojure.string :as str]
            [jsonista.core :as json]))

(def records-store (atom []))

(defn records-sorted-by [criterion]
  (->> @records-store
    (sort-by criterion)
    (json/write-value-as-string)))

(defn records-json [recs]
  {:status 200
   :headers {"Content-Type" "application/json"}
   :body recs})

(defn add-record [body type]
  (let [newrec (rec/line->record body (rec/suffix->delimiter type))]
    (swap! records-store conj newrec)
    {:status 200
     :content-type "application/jsonplain"
     :body (json/write-value-as-string newrec)}))
  
(defn records-by-color [req]
  (records-json (records-sorted-by :favorite-color)))
(defn records-by-birthdate [req]
  (records-json (records-sorted-by :birthdate)))
(defn records-by-name [req]
  (records-json (records-sorted-by (juxt :last-name :first-name))))

(defroutes app-routes
  (POST "/records" [type :as {body :body}]
    (add-record (slurp body) (keyword type)))
  (GET "/records/color" [] records-by-color)
  (GET "/records/birthdate" [] records-by-birthdate)
  (GET "/records/name" [] records-by-name)
  (route/not-found "error, no route"))

(defn -main
  "Read files named in command line args, sort and print"
  [& args]
  (swap! records-store conj (mapcat rec/file->records args))
  (server/run-server (wrap-defaults #'app-routes api-defaults)
                     {:port 3000})
  (println (str "Running API on port 3000")))
